sdks/python/apache_beam/ml/inference/tensorrt_inference.py

```diff
@@ -0,0 +1,294 @@
+                                              TensorRTEngine]):
+  def __init__(self, min_batch_size: int, max_batch_size: int, **kwargs):
+    """Implementation of the ModelHandler interface for TensorRT.
+
+    Example Usage::
+
+      pcoll | RunInference(
+          TensorRTEngineHandlerNumPy(
+            min_batch_size=1,
+            max_batch_size=1,
+            engine_path="my_uri"))
+
+    Args:
+      min_batch_size: minimum accepted batch size.
+      max_batch_size: maximum accepted batch size.
+      kwargs: Additional arguments like 'engine_path' and 'onnx_path' are
+        currently supported.
+
+    See https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/
+    for details
+    """
+    self.min_batch_size = min_batch_size
+    self.max_batch_size = max_batch_size
+    if 'engine_path' in kwargs:
+      self.engine_path = kwargs.get('engine_path')
+    elif 'onnx_path' in kwargs:
+      self.onnx_path = kwargs.get('onnx_path')
+
+  def batch_elements_kwargs(self):
+    """Sets min_batch_size and max_batch_size of a TensorRT engine."""
+    return {
+        'min_batch_size': self.min_batch_size,
+        'max_batch_size': self.max_batch_size
+    }
+
+  def load_model(self) -> TensorRTEngine:
+    """Loads and initializes a TensorRT engine for processing."""
+    engine = _load_engine(self.engine_path)
+    return TensorRTEngine(engine)
+
+  def load_onnx(self) -> Tuple[trt.INetworkDefinition, trt.Builder]:
+    """Loads and parses an onnx model for processing."""
+    return _load_onnx(self.onnx_path)
+
+  def build_engine(
+      self, network: trt.INetworkDefinition,
+      builder: trt.Builder) -> TensorRTEngine:
+    """Build an engine according to parsed/created network."""
+    engine = _build_engine(network, builder)
+    return TensorRTEngine(engine)
+
+  def run_inference(
+      self,
+      batch: Sequence[np.ndarray],
+      engine: TensorRTEngine,
+      inference_args: Optional[Dict[str, Any]] = None
+  ) -> Iterable[PredictionResult]:
+    """
+    Runs inferences on a batch of Tensors and returns an Iterable of
+    TensorRT Predictions.
+
+    Args:
+      batch: A np.ndarray or a np.ndarray that represents a concatenation
+        of multiple arrays as a batch.
+      engine: A TensorRT engine.
+      inference_args: Any additional arguments for an inference
+        that are not applicable to TensorRT.
+
+    Returns:
+      An Iterable of type PredictionResult.
+    """
+    from cuda import cuda
+    _validate_inference_args(inference_args)
+    (
+        engine,
+        context,
+        inputs,
+        outputs,
+        gpu_allocations,
+        cpu_allocations,
+        stream) = engine.get_engine_attrs()
+    # Process I/O and execute the network
+    _assign_or_fail(
+        cuda.cuMemcpyHtoDAsync(
+            inputs[0]['allocation'],
+            np.ascontiguousarray(batch),
+            inputs[0]['size'],
+            stream))
+    context.execute_async_v2(gpu_allocations, stream)
```

Comment:
Given that we use async, is this a blocking call?

Reply:
@pranavm-nvidia Maybe you know better how to answer this.

Reply:
I don't think it is a blocking call

Reply:
`execute_async_v2` is a non-blocking call

Reply:
I see. For my education, what mechanism ensures that the predictions are computed by the time the function call returns? Thanks.

Reply:
I guess doing it at the end of every single prediction call is not a good idea?

Reply:
In this case, synchronizing after each prediction call makes sense. 

If instead, e.g., you synchronized after every other call, I don't see how you'd access the prediction results from the first call, as they'd be overwritten by the second call.

Reply:
Yeah makes sense. Ok, I'll add stream sync at the end of the call

Reply:
Added [stream sync](https://github.com/apache/beam/pull/22131/commits/5d52984f5db83edce7a8e1cc54fe65ee019f3e03#diff-145530e8e39084aed7792a4d3b8104258ea1635db466ab48a58f168bce10dde3R269) before copying data to host.

Reply:
The device-to-host copy is also asynchronous, so the stream sync would have to be after that.

Reply:
Fixed. Thank you @pranavm-nvidia

Reply:
Would using `execute_v2()` not achienve the same effect as using the async version + doing a sync later?

Reply:
Nearly, except that it wouldn't allow us to overlap the memcpys with inference. It probably makes no difference for single input/output scenarios, but it could help if you plan to eventually extend the interface to multiple inputs/outputs.

Reply:
It already supports multiple outputs, at least in case of object detection it will be able to handle 4 outputs: boxes, classes, scores and number of predictions per image. The only thing left is to add support for multiple inputs, probably in future.

Reply:
@tvalentyn Do you think we can close it?